packages/checkout-storefront/src/sections/BillingAddressSection/useSetBillingSameAsShipping.ts

```diff
@@ -0,0 +1,64 @@
+import { Address, AddressFormData } from "@/checkout-storefront/components/AddressForm/types";
+import { useCheckout } from "@/checkout-storefront/hooks";
+import { getAddressFormDataFromAddress, isMatchingAddress } from "@/checkout-storefront/lib/utils";
+import { useCallback, useEffect, useRef, useState } from "react";
+
+interface UseSetBillingSameAsShippingProps {
+  handleSubmit: (address: AddressFormData) => void;
+}
+
+export const useSetBillingSameAsShipping = ({ handleSubmit }: UseSetBillingSameAsShippingProps) => {
+  const { checkout } = useCheckout();
+  const { billingAddress, shippingAddress } = checkout;
+
+  const [passDefaultFormDataAddress, setPassDefaultFormDataAddress] = useState<boolean>(
+    !!billingAddress
+  );
+
+  const hasBillingSameAsShipping = isMatchingAddress(shippingAddress, billingAddress);
+
+  const [isBillingSameAsShipping, setIsBillingSameAsShipping] = useState<boolean>(
+    checkout?.isShippingRequired ? !billingAddress || hasBillingSameAsShipping : false
+  );
+  const isBillingSameAsShippingRef = useRef<boolean>(isBillingSameAsShipping);
+  const shippingAddressRef = useRef<Address>(shippingAddress);
+
+  const setBillingSameAsShipping = useCallback(async () => {
+    if (!hasBillingSameAsShipping && shippingAddress) {
+      void handleSubmit({
+        ...getAddressFormDataFromAddress(shippingAddress),
+        autoSave: true,
+      });
+    }
+  }, [handleSubmit, hasBillingSameAsShipping, shippingAddress]);
+
+  useEffect(() => {
+    const billingSetDifferentThanShipping =
+      !isBillingSameAsShipping && isBillingSameAsShippingRef.current;
+
+    if (billingSetDifferentThanShipping) {
+      setPassDefaultFormDataAddress(false);
+      isBillingSameAsShippingRef.current = isBillingSameAsShipping;
+    }
+  }, [isBillingSameAsShipping]);
+
+  useEffect(() => {
+    if (!isBillingSameAsShipping) {
+      return;
+    }
+
+    const billingSetSameAsShipping = isBillingSameAsShipping && !isBillingSameAsShippingRef.current;
+
+    const hasShippingAddressChanged =
+      shippingAddress && shippingAddress !== shippingAddressRef.current;
+
+    if (hasShippingAddressChanged || billingSetSameAsShipping) {
+      void setBillingSameAsShipping();
+      shippingAddressRef.current = shippingAddress;
+      isBillingSameAsShippingRef.current = isBillingSameAsShipping;
+      return;
+    }
+  }, [shippingAddress, isBillingSameAsShipping, setBillingSameAsShipping]);
+
+  return { isBillingSameAsShipping, setIsBillingSameAsShipping, passDefaultFormDataAddress };
+};
```

Comment:
Yes and no. Same reason useSubmit doesn't have tests for now - it uses `useCheckout` which uses `useAuthState` from saleor sdk. We're currently stuck with Saleor SDK that will not be maintained from now on which uses apollo client underneath and requires us to use `SaleorProvider` which in turn is really painful to mock (especially since SDK doesn't provide a way for it). Since after merging this pr all we've got left to replace of SDK functionality is authentication we'll do it very soon and then add tests for both useSubmit and useSetBillingSameAsShipping